### PR TITLE
[Issue 15768] Make std.stdio.File refcount shared(uint)

### PR DESCRIPTION
**Pull request by JackStouffer**
This is a band-aid, not a real fix for the underlying problem that stdout and stdin need to be `shared`.

With 20 runs of the sample program in the report, I was not able to reproduce the race condition after this fix.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Ping @schveiguy @atilaneves @mleise

File: std/stdio.d [hunk 1]
```diff
@@ -651,15 +652,15 @@ Detaches from the underlying file. If the sole owner, calls $(D close).
...
 Throws: $(D ErrnoException) on failure if closing the file.
   */
-    void detach() @safe
+    void detach() @trusted
     {
         if (!_p) return;
-        if (_p.refs == 1)
+        if (atomicLoad(_p.refs) == 1)
             close();
         else
```
Comment (schveiguy) [inline std/stdio.d]: No, this is not thread-correct.

Assume 2 threads are detaching. Both execute detach simultaneously, both atomically load 2 as the reference count. Both will simply decrement the reference count, resulting in RC of 0, but no destruction. The resource is then leaked.

The correct way to do this is to atomically decrement, and check for 0.

**Comment by schveiguy** [comment]
I don't know how to begin fixing this error. std.stdio.File is so borked in terms of sharing that I think we may need a drastic redesign.

I do like the idea of having the atomic operations on the reference count. That should be the most performant way to go about fixing.

File: std/stdio.d (hunk 1 above)
Comment (JackStouffer) [inline std/stdio.d]: > The correct way to do this is to atomically decrement, and check for 0.

I don't think that's possible, as close expects the refcount to be 1, and close can be called by the programmer. 

The only option I see now is locking in detach.

Comment (schveiguy) [inline std/stdio.d]: I think a rearrangement of what is done is in order:

1. You need a private function that closes the file without detaching.
2. close needs to call private func, and then detach
3. All RC should be done in detach.

**Comment by MartinNowak** [comment]
Can we do the atomic ref counting only in a `this(this) shared` overload?
Is the rest of File usable as shared?

**Comment by MartinNowak** [comment]
You commit message should say fix Issue 12345, in order to be correctly referenced, this is indicated by the check mark in dlang-bot's comment.

**Comment by andralex** [comment]
This is not the best solution here - we're paying the cost of shared refcount for the sake of the three special handles.

There are several other possibilities here. A simple one is to simply check whether `this` is one of the special three files, and if it isn't just use non-atomic operations.

But the real matter here is the use of `__gshared`, which makes `stdout` unusable in safe code - something we definitely need to fix. At the time `std.stdio` was defined, `shared` was more buggy than now, so perhaps this is the right time to investigate what keeps us from using `shared(File)` properly.

That has other benefits as well: for non-`shared` `File`s we can just use the unlocked I/O primitives, which are faster.

**Comment by JackStouffer** [comment]
I'm closing this for now. The right solution is to write a `shared(File)` implementation that's thread safe. That will require a lot of work that I'm not willing to take on at this moment.

**Pull request closed without merging**